File: job/jobs.go
// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
// See License.txt for license information.

package job

import (
	"sync"

	l4g "github.com/alecthomas/log4go"
	"github.com/mattermost/platform/store"
)

type Jobs struct {
	startOnce sync.Once

	DataRetention  Job
	SearchIndexing Job
}

func InitJobs(s store.Store) *Jobs {
	return &Jobs{
		DataRetention:  MakeTestJob(s, "DataRetention"),
		SearchIndexing: MakeTestJob(s, "SearchIndexing"),
	}
}

func (jobs *Jobs) StartAll() *Jobs {
	l4g.Info("Starting jobs")

	jobs.startOnce.Do(func() {
		go jobs.DataRetention.Run()
		go jobs.SearchIndexing.Run()
	})

	return jobs
}

func (jobs *Jobs) StopAll() *Jobs {
	l4g.Info("Stopping jobs")

	jobs.DataRetention.Stop()
	jobs.SearchIndexing.Stop()

	l4g.Info("Stopped jobs")

	return jobs
}

File: webapp/channels/src/packages/mattermost-redux/src/selectors/entities/jobs.ts
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {createSelector} from 'mattermost-redux/selectors/create_selector';

import type {JobType, Job, JobsByType} from '@mattermost/types/jobs';
import type {GlobalState} from '@mattermost/types/store';
import type {IDMappedObjects} from '@mattermost/types/utilities';

export function getAllJobs(state: GlobalState): IDMappedObjects<Job> {
    return state.entities.jobs.jobs;
}

export function getJobsByType(state: GlobalState): JobsByType {
    return state.entities.jobs.jobsByTypeList;
}

export function makeGetJobsByType(type: JobType): (state: GlobalState) => Job[] {
    return createSelector(
        'makeGetJobsByType',
        getJobsByType,
        (jobsByType) => {
            return jobsByType[type] || [];
        },
    );
}
